### PR TITLE
Allow fractional fee multipliers (RIPD-626)

### PR DESCRIPTION
**Pull request by ximinez**
Also move `mulDiv` out of STAmount, and make it more "0-safe".

Reviewers: @scottschurr, @vinniefalco. Also mentioning @rec as the initial requester on the Jira ticket.

### CODE REVIEW

File: src/ripple/basics/mulDiv.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_MULDIV_H_INCLUDED
+#define RIPPLE_BASICS_MULDIV_H_INCLUDED
+
+namespace ripple
```
Comment (vinniefalco) [inline src/ripple/basics/mulDiv.h]: missing `<cstdint>`

Comment (ximinez) [inline src/ripple/basics/mulDiv.h]: Thanks. Fixed. Unity and nounity builds both worked in Windows, so I didn't even think about it.

**Comment by scottschurr** [comment]
I couldn't find a single thing to gripe about.  Looks good to me.
:+1:

**Comment by scottschurr** [comment]
Oh, and I forgot to mention, nice tests.

**Comment by ximinez** [comment]
Thanks!

On Tue, Dec 8, 2015 at 5:49 PM Scott Schurr <notifications@github.com>
wrote:

> Oh, and I forgot to mention, nice tests.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ripple/rippled/pull/1423#issuecomment-163045846>.
>
-- 

*Edward Hennis*Senior Software Engineer | Ripple
+352.327.8769
@EdwardHennis
PGP: F2198E3A / 5294 9650 7465 B591 A982  831C D835 08C0 F219 8E3A
ed@ripple.com | ripple.com

**Comment by ximinez** [comment]
Rebased against 0.30.1-b12.

**Comment by ximinez** [comment]
Rebased against 0.30.1-b13.

**Comment by ximinez** [comment]
Pulled in changes from #1461 in the hopes of fixing CI failures.

File: src/ripple/basics/mulDiv.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+#include <cstdint>
+
+namespace ripple
+{
+
+/**
+    A utility function to compute (value)*(mul)/(div) while avoiding
+    overflow but keeping precision.
+*/
```
Comment (vinniefalco) [inline src/ripple/basics/mulDiv.h]: Comment could be better. The first line needs to be a complete sentence. And whats a "utility function?" I don't see that anywhere in the C++ standard. Consider:
```
/** Return value*mul/div accurately.
    Computes the result of the multiplication and division in
    a single step, avoiding overflow and retaining precision.
    Throws:
        ...
*/
```

File: src/ripple/basics/mulDiv.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+std::uint64_t
+mulDiv(std::uint64_t value, std::uint64_t mul, std::uint64_t div);
+
+/**
+    A utility function to compute (value)*(mul)/(div) while avoiding
+    overflow but keeping precision. Will return the max uint64_t
+    value if mulDiv would overflow anyway.
```
Comment (vinniefalco) [inline src/ripple/basics/mulDiv.h]: Same comment regarding the Javadoc

File: src/ripple/basics/mulDiv.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+*/
+std::uint64_t
+mulDivNoThrow(std::uint64_t value, std::uint64_t mul, std::uint64_t div);
+
+template <class T1, class T2>
+void lowestTerms(T1& a,  T2& b)
```
Comment (vinniefalco) [inline src/ripple/basics/mulDiv.h]: Consider constraining the function (using `std::enable_if`) where `U` and `V` are integral:
```
template <class U, class V>
...
```

**Comment by ximinez** [comment]
Rebased onto 0.30.1-b14.

File: src/ripple/basics/mulDiv.h (hunk 2 above)
Comment (ximinez) [inline src/ripple/basics/mulDiv.h]: Fixed

File: src/ripple/basics/mulDiv.h (hunk 3 above)
Comment (ximinez) [inline src/ripple/basics/mulDiv.h]: Fixed

File: src/ripple/basics/mulDiv.h (hunk 4 above)
Comment (ximinez) [inline src/ripple/basics/mulDiv.h]: Done.

File: src/ripple/basics/impl/mulDiv.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    lowestTerms(value, div);
+    lowestTerms(mul, div);
+
+    if (value < mul)
+        std::swap(value, mul);
+    const auto max = std::numeric_limits<std::uint64_t>::max();
+    if (value > max / mul)
+    {
+        value /= div;
+        if (value > max / mul)
```
Comment (vinniefalco) [inline src/ripple/basics/impl/mulDiv.cpp]: `max / mul` should be stored in a local variable and reused?

**Comment by vinniefalco** [comment]
Good enough for government work!
:+1:

**Comment by ximinez** [comment]
@vinniefalco I pushed another commit with a couple of small fixes, adding unit tests, and cleaning up the `mulDivNoThrow` issues a bit. Check it out if you'd like, though I'm going to mark this as "Passed" anyway.

CircleCI and Travis are failing because llvm.org is currently down. I have verified as much as possible locally that MSVC, clang, and gcc builds work and pass tests.

**Pull request closed without merging**